### PR TITLE
Add compact binary specification API-1759

### PR DESCRIPTION
**Pull request by srknzl**
Publication of https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/3507487040/Compact+Wire-level+Format+Design+Document with some changes. The purpose is to describe our binary compact format to public. 

I refactored the document so that we don't use `we` and `will` words. I also removed the sections about discussion and brain storming. The document should be format and describe what is the binary format as of now. I used grammarly to do a grammar check after that.

### CODE REVIEW

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+YYYY-MM-DD from -999999999-01-1 to 999999999-12-31 +
+int32: year +
+ int8: month +
+int8: dayOfMonth
+|no(since it is nullable)
+|timestamp
+|LocalDateTime
+|optional<hazelcast:client:local_date_time>
+|HLocalDateTime?
+|typing.Optional[datetime.datetime]
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:138]: for python, can we get rid of `typing.` prefix, just like the table we have in the https://docs.hazelcast.com/hazelcast/latest/serialization/compact-serialization#supported-types?

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+|bool[]?
+|typing.Optional[typing.List[bool]]
+|boolean[] \| null
+|[]bool
+|
+|Array of booleans
+|no
+|int8[]
+|byte[]
+|boost::optional<std::vector<int8_t>>
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:194]: sometimes we use `boost::optional` sometimes just the `optional`. I think we should stick to one for C++

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+|float[]?
+|typing.Optional[typing.List[float]]
+|number[] \| null
+|[]float32
+|
+|Array of float32s
+|no
+|float64[]
+|double[] 
+|oost::optional<std::vector<double>>
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:244]: probably a copy paste error

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+|Array of nullable float64
+|no
+|===
+
+
+=== Type Ids
+
+Each type supported in the wire format has its type id. The type ids are used while constructing the schemas, performing type checks for user access to fields, and are exposed to the user as a public API.
+
+To distinguish different types supported in the wire-level format, a new enum called `FieldKind` is introduced which has different ids for different types compared to the `FieldType` enum.
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:471]: I think we don't need this information regarding the `FieldKind` and `FieldType` enums. They have nothing to do with the specification

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+
+The `FieldType` enum has different ways to represent ids for types and their arrays and has some unfortunate public methods which would not fit the new format. (for example `FieldType#DATE` returns `getTypeSize` wrong for the new format). It is desired to have a new enum that does not have those problems and is more open to type additions in the future.
+
+Note that, the new `FieldKind` enum with those new ids will be used in the `GenericRecord` API for both `Portable` and the new format. `FieldType` and `Portable` will be deprecated in the future.
+
+=== Unsigned Integers
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:477]: We don't have support for unsigned integers right now

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+There is support for unsigned integer types. The representation of unsigned integers is done with the smallest signed integer type that can represent it (except for `u64`, since there is no such type). So, basically,
+
+* `u8` is represented by → `i16`
+* `u16` is represented by → `i32`
+* `u32` is represented by → `i64`
+* `u64` is represented by → `BigInteger`
+
+=== Enums
+
+There is no support for enums in the wire-level format. The representation of enums is left to the user.
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:488]: Then, shouldn't we remove this from our reference?

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 7]
```diff
@@ -0,0 +1,719 @@
...
+* `u32` is represented by → `i64`
+* `u64` is represented by → `BigInteger`
+
+=== Enums
+
+There is no support for enums in the wire-level format. The representation of enums is left to the user. 
+
+=== Nullable Primitives
+
+There is support for nullable primitive types as they can be useful in SQL and might play nicely with languages that have a concept to represent them. For example, a C# user might use `Nullable<PrimitiveType>`, or Java users might use the class representation of primitive types.
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:492]: Should we talk about the languages in this reference? I thought it would be something language agnostic. I believe, instead of saying that we do it for this and that language, we should just specify how we do it

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 8]
```diff
@@ -0,0 +1,719 @@
...
+|Data length |i32 |Length of the DATA SECTION below.
+|===
+
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Description
+|Fixed-size Fields | Offsets of these fields will be deduced from the schema
+|Variable-size Fields |
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:527]: No description for this row?

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 9]
```diff
@@ -0,0 +1,719 @@
...
+|Fixed-size Fields | Offsets of these fields will be deduced from the schema
+|Variable-size Fields | 
+|===
+
+=== Offsets Section
+
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Variable-Size FieldOffset index 0 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
```
Comment (mdumandag) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:535]: I believe we should either duplicate the description to all rows, or remove that column altogether here and explain the null case below

File: docs/modules/architecture/partials/nav.adoc [hunk 10]
```diff
@@ -6,7 +6,7 @@
 ** xref:architecture:event-time-processing.adoc[]
 ** xref:architecture:sliding-window.adoc[]
 ** xref:architecture:in-memory-storage.adoc[]
-
+** xref:architecture:compact-binary-specification.adoc[]
```
Comment (mdumandag) [inline docs/modules/architecture/partials/nav.adoc:9]: IMO, this is not really related to architecture

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 11]
```diff
@@ -0,0 +1,719 @@
+= Compact Serialization Binary Specification
+:description: This document describes the binary specification of compact serialization. You will learn what serialized data look like in binary.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/architecture/pages/compact-binary-specification.adoc:2]: Changed to `This document will describe what serialized data look like in binary`

File: docs/modules/architecture/pages/compact-binary-specification.adoc (hunk 5 above)
Comment (srknzl) [reply docs/modules/architecture/pages/compact-binary-specification.adoc:477]: removing the section.

File: docs/modules/architecture/pages/compact-binary-specification.adoc (hunk 8 above)
Comment (srknzl) [reply docs/modules/architecture/pages/compact-binary-specification.adoc:527]: There wasn't a description basically. 
image
Does this description apply to both? https://hazelcast.atlassian.net/wiki/spaces/IMDG/pages/3507487040/Compact+Wire-level+Format+Design+Document#Var-Size-Objects

File: docs/modules/architecture/partials/nav.adoc (hunk 10 above)
Comment (srknzl) [reply docs/modules/architecture/partials/nav.adoc:9]: Isn't it a design document regarding serialization? So I thought it is related to serialization architecture. Where do you think we should put the document instead?

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 12]
```diff
@@ -0,0 +1,706 @@
...
+|[]*float64
+|
+|Array of nullable float64
+|no
+|===
+
+
+=== Type Ids
+
+Each type supported in the wire format has its type id. The type ids are used while constructing the schemas, performing type checks for user access to fields, and are exposed to the user as a public API.
```
Comment (Serdaro) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:477]: === Type IDs

Each type supported in the wire format has its type ID. The type IDs are used while constructing the schemas, performing type checks when accessing to fields, and are exposed as a public API.

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 13]
```diff
@@ -0,0 +1,706 @@
...
+
+=== Nullable Primitives
+
+They are implemented as variable-sized types. The `null` values of such types are represented exactly as `null` variable-sized fields, with the offset of `-1` and no data.
+
+== Header
+
+The partition hash and the type id are common for all serialization methods supported by Hazelcast. Therefore, the new format is no exception and every serialized object has a header in addition to the payload on the wire.
```
Comment (Serdaro) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:485]: ... the type ID are ...

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 14]
```diff
@@ -0,0 +1,706 @@
...
+== Schema ID
+
+We are using 64bit https://en.wikipedia.org/wiki/Rabin_fingerprint[Rabin fingerprint] to create [.inline-comment-marker]#a schema id#.
+
+Rabin fingerprint is chosen mostly because it is recommended in Avro’s documents as follows +
+https://avro.apache.org/docs/current/spec.html#schema_primitive
+
+[source,text]
+----
+At the opposite extreme, the smallest fingerprint recommended is a 64-bit Rabin fingerprint. Below, there is a provided pseudo-code for this algorithm that can be easily translated into any programming language. 64-bit fingerprints should guarantee uniqueness for schema caches of up to a million entries (for such a cache, the chance of a collision is 3E-8). It is not recommended to use shorter fingerprints, as the chances of collisions are too high (for example, with 32-bit fingerprints, a cache with as few as 100,000 schemas has a 50% chance of having a collision).
```
Comment (Serdaro) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:604]: What does "opposite extreme" mean here in this context?

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 15]
```diff
@@ -0,0 +1,706 @@
...
+|Item 1 offset | u8/u16/i32
+|... | ...
+|Item n offset | u8/u16/i32
+|===
+
+
+---
+
+An array can contain only a single type of item.
+In the case of `Compact[]` all the items must have the same schema, i.e their schema id must be equal.
```
Comment (Serdaro) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:690]: In the case of `Compact[]`, all the items must have the same schema, i.e, their schema ID must be equal.

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 16]
```diff
@@ -0,0 +1,706 @@
...
+
+* Data Length <= `254`, offsets are `u8` (`255` is reserved for `null`)
+* Data Length <= `65534`, offsets are `u16` (`65535` is reserved for `null`)
+* Otherwise, offsets are `i32`.
+
+Items can be `null`. The corresponding offset will be set to `-1` in that case.
+
+== Nullable Values
+
+Fixed-size fields will always be on the binary and take up space. On the other hand, when variable-size fields are set to `null`, their offset will be set to `-1` in the binary, and no further data will be written.
```
Comment (Serdaro) [inline docs/modules/architecture/pages/compact-binary-specification.adoc:706]: Fixed-size fields will be always on the binary and take up space. On the other hand, when variable-size fields are set to `null`, their offset will be set to `-1` in the binary, and no further data will be written.

**Review by Serdaro — COMMENTED** [review]
Minor suggestions.

File: docs/modules/architecture/pages/compact-binary-specification.adoc [hunk 17]
```diff
@@ -0,0 +1,706 @@
...
+[cols="1,1,1"]
+|===
+|Name |Type |Description
+|Partition hash |i32 |`BIG_ENDIAN` integer, used for key objects. Not applicable to value objects.
+|Type id |i32 |`BIG_ENDIAN` integer that determines the serializer to be used. -55 for compact.
+|===
+
+== Var-Size Objects
+
+In this section, how a user-defined type is represented at the wire level is described. Consists of `Header`, `Data`, and `Offsets` sections in this order.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/architecture/pages/compact-binary-specification.adoc:496]: Applied as `The binary consists of` because we talk about the binary, not the type.

File: docs/modules/architecture/pages/compact-binary-specification.adoc (hunk 14 above)
Comment (srknzl) [reply docs/modules/architecture/pages/compact-binary-specification.adoc:604]: The paragraph is from https://avro.apache.org/docs/1.11.1/specification/_print/ . It means high number of bits fingerprint algorithms as opposed to low number of bits. (e.g 256bit instead of 64)


What do you think we should change here?

File: docs/modules/architecture/partials/nav.adoc (hunk 10 above)
Comment (mdumandag) [reply docs/modules/architecture/partials/nav.adoc:9]: I don't think serialization is part of the architecture. I believe we should have a separate section in the reference tab called sterilization, at put this under it

Comment (srknzl) [reply docs/modules/architecture/partials/nav.adoc:9]: As I said in slack as well, I think we can put it there as well. I will change the location of the document. You can keep the comment unresolved till I do it as a reminder to me.

Comment (Serdaro) [reply docs/modules/architecture/partials/nav.adoc:9]: I was simply thinking to put it as a section under Develop Solutions -> Serialization -> Compact Serialization.

If not, there should be a cross-reference to this spec API from the introduction part of the Develop Solutions -> Serialization -> Compact Serialization section.

Comment (srknzl) [reply docs/modules/architecture/partials/nav.adoc:9]: @Serdaro Isn't the document too long for being a section in a page ? I think it should be a separate page. I think cross reference makes more sense for this reason. Where should be the page's location in this case? I think a page directly under "Reference" makes sense. (I am also ok about putting it under architecture obviously)

**Review by Serdaro — APPROVED** [review]

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 18]
```diff
@@ -0,0 +1,706 @@
+= Compact Serialization Binary Specification
+:description: This topic describes the binary specification of Compact serialization, showing how serialized data look like in binary format.
```
Comment (mdumandag) [inline docs/modules/ROOT/pages/compact-binary-specification.adoc:2]: Maybe it is just me, but I feel like the `topic` is not a good word here. Maybe something like `document`, or `page` can be used instead?

Or maybe we can even change the structure of the sentence altogether and say something like

`This specification describes the binary format of the Compact serialization`.

**Review by mdumandag — APPROVED** [review]

File: docs/modules/ROOT/pages/compact-binary-specification.adoc (hunk 18 above)
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:2]: It is `This document describes the binary format of the Compact serialization.` now.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 19]
```diff
@@ -0,0 +1,706 @@
...
+
+== Introduction
+
+Every serialized user object consists of a header and data. There is also a schema that is separate from the serialized object. Schemas and serialized user objects are linked using the schema ID written in the data section.
+
+== Data Types
+
+The table below shows the data types that are supported by the Compact serialization.
+
+You can configure the endianness for primitives; they are written in `BIG_ENDIAN` by default. This configuration is applied to user types and all the primitives used in the format. These include the size of arrays, number of items, length of data, schema ID and offsets. A new endianness configuration does not exist for Compact serialization; it is inherited from the xref:serialization:serialization-configuration.adoc#configuration-options[existing configuration].
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:14]: Well, thanks it is more clear now. We wanted to mean what you wrote but it was a little indirect.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 20]
```diff
@@ -0,0 +1,706 @@
...
+
+Each type supported in the wire format has its type ID. The type IDs are used while constructing the schemas, performing type checks when accessing to fields, and are exposed as a public API.
+
+=== Nullable Primitives
+
+They are implemented as variable-sized types. The `null` values of such types are represented exactly as `null` variable-sized fields, with the offset of `-1` and no data.
+
+== Header
+
+The partition hash and the type ID are common for all serialization methods supported by Hazelcast. Therefore, the new format is no exception and every serialized object has a header in addition to the payload on the wire.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:485]: Applied

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 21]
```diff
@@ -0,0 +1,706 @@
...
+=== Data Section
+
+[cols="1,1"]
+|===
+|Name |Description
+|Fixed-size Fields | Offsets of these fields will be deduced from the schema.
+|Variable-size Fields | 
+|===
+
+=== Offsets Section
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:517]: What is the reasoning behind removing section ?

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 22]
```diff
@@ -0,0 +1,706 @@
...
+
+[cols="1,1,3"]
+|===
+|Name |Type |Description
+|Variable-Size FieldOffset index 0 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:522]: Left "the Compact Schema" I newly added in the beginning capitalized because it is a term.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 23]
```diff
@@ -0,0 +1,706 @@
...
+|Variable-Size FieldOffset index 1 |u8/u16/i32 |The index of a field offset is written in the Schema. Offsets of variable length fields. -1 for null
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:523]: Replaced all of them by finding them by capitalization sensitive searching.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 24]
```diff
@@ -0,0 +1,706 @@
...
+
+* Data Length <= `254`, offsets are `u8` (`255` is reserved for `null`)
+* Data Length <= `65534`, offsets are `u16` (`65535` is reserved for `null`)
+* Otherwise, offsets are `i32`.
+
+Length is written before offsets so that the binary can be skipped even when the schema cannot be found.
+
+A Variable-Size FieldOffset is `-1` if a Variable-Size field is `null`.
+
+Fixed-Size Fields cannot be `null`.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:544]: The information refers to table row names so I think it is better this way.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 25]
```diff
@@ -0,0 +1,706 @@
...
+
+== Fixed-Size Fields
+
+The fixed-size fields are written right after the `Length` field consecutively. They are accessed via `offset` written in the Schema.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:548]: Partially applied, the offset is not given in the schema it is calculated from. Made second sentence 'Fixed-size fields are accessed via their offsets calculated from the schema'

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 26]
```diff
@@ -0,0 +1,706 @@
...
+
+On the schema, the offset for a fixed-size field is determined as follows:
+
+* The first field always starts from offset 0.
+* Fields are ordered by their size in descending order.
+* When sizes are the same the fields are ordered by field name.
+* Each offset is calculated by adding the size of the last field to the last offset.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:555]: The first point states that: "The first field always starts from offset 0.". Isn't this enough?

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 27]
```diff
@@ -0,0 +1,706 @@
...
+
+The only exception to the above rule is boolean fields. Since up to 8 booleans can be packed into a single byte, they are treated specially, and extra information is stored in the schema (nothing extra on the data) for the bit index of the boolean fields. Boolean fields are written at the end of the fixed-size fields.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:557]: Added some more detail in relation to the rule and booleans.

**Comment by srknzl** [comment]
@llivadas-mdsol I haven't checked all your comment but maybe half of them. Will continue working on them. FYI

File: docs/modules/ROOT/pages/compact-binary-specification.adoc (hunk 21 above)
[the remark this answers is not in the document]
Comment (llivadas-mdsol) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc]: @srknzl The reason is twofold:
1. To avoid repetition.
2. The document header size and style (bold) mark it out as a section.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc (hunk 25 above)
[the remark this answers is not in the document]
Comment (llivadas-mdsol) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc]: @srknzl Agreed. Please also add "that are" between "offsets" and "calculated", and a full stop at the end of the sentence. Thank you.

[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc]: there was a full stop, I added "that are" (btw if you see my comments right after I write them, I may not have pushed the changes yet, it will be eventually there though)

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 28]
```diff
@@ -0,0 +1,706 @@
...
+
+== Variable-Size Fields
+
+The offsets of variable-size fields are written at the end in the alphabetical order of the field names. To read a variable-size field from the data, one should read the index of the offset from the schema. Then, the related index is read from the end of the data to get the offset. The variable-size field can be read using this offset.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:574]: Isn't what you wrote is the same except the first sentence removed? To be able to refer to the offset, we mention the offset it in the first sentence.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 29]
```diff
@@ -0,0 +1,706 @@
...
+|name of field 1 |string
+|type ID of field 1 |i32
+|... |...
+|name of field n |string
+|type ID of field n |i32
+|===
+
+When writing a schema to the wire, fields are ordered according to their names so that the same structure results in the same byte representation and produces the same schema ID.
+
+The offsets and indexes are also decided on the ordered fields. The smaller-sized fields come first. The order is by name within the same size fields.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:600]: I agree that whole part was complex. Made it clearer hopefully.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 30]
```diff
@@ -0,0 +1,706 @@
...
+    for (int j = 0; j < 8; j++)
+      fp = (fp >>> 1) ^ (EMPTY & -(fp & 1L));
+    FP_TABLE[i] = fp;
+  }
+}
+----
+
+== Arrays
+
+Arrays of fix-sized items can not have `null` items. On the other hand, arrays of variable-size items may contain `null` items.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:650]: made it "fixed-size"

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 31]
```diff
@@ -0,0 +1,706 @@
...
+== Schema ID
+
+We are using 64bit https://en.wikipedia.org/wiki/Rabin_fingerprint[Rabin fingerprint] to create a schema ID.
+
+Rabin fingerprint is chosen mostly because it is recommended in Avro’s 
+https://avro.apache.org/docs/1.11.1/specification/_print/#schema-fingerprints[documentation^] as follows.
+
+[quote, Apache Avro Specification]
+----
+At the opposite extreme, the smallest fingerprint recommended is a 64-bit Rabin fingerprint. Below, there is a provided pseudo-code for this algorithm that can be easily translated into any programming language. 64-bit fingerprints should guarantee uniqueness for schema caches of up to a million entries (for such a cache, the chance of a collision is 3E-8). It is not recommended to use shorter fingerprints, as the chances of collisions are too high (for example, with 32-bit fingerprints, a cache with as few as 100,000 schemas has a 50% chance of having a collision).
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:617]: The quote describes that even with a million entries change of a collision is very low. So, we think that description is enough and we won't change number of bits in the near future. Let me explicitly state that.

File: docs/modules/ROOT/pages/compact-binary-specification.adoc [hunk 32]
```diff
@@ -0,0 +1,706 @@
...
+
+* Data Length <= `254`, offsets are `u8` (`255` is reserved for `null`)
+* Data Length <= `65534`, offsets are `u16` (`65535` is reserved for `null`)
+* Otherwise, offsets are `i32`.
+
+Items can be `null`. The corresponding offset will be set to `-1` in that case.
+
+== Nullable Values
+
+Fixed-size fields will be always on the binary and take up space. On the other hand, when variable-size fields are set to `null`, their offset will be set to `-1` in the binary, and no further data will be written.
```
[the remark this answers is not in the document]
Comment (srknzl) [reply docs/modules/ROOT/pages/compact-binary-specification.adoc:719]: Tried to make this more clear by adding more description and trying to remove repetition.

**Comment by srknzl** [comment]
@llivadas-mdsol I have just checked all of your comments. If a put a 👍🏻  it means I applied your comment. Feel free to go over your comments again to check if there's anything I overlooked.

**Review by llivadas-mdsol — APPROVED** [review]
LGTM. Thank you.

**Review by llivadas-mdsol — APPROVED** [review]
LGTM. Thank you.

**Comment by srknzl** [comment]
@llivadas-mdsol Thank you, too.
